### PR TITLE
Handle segments outside retention time gracefully during creation

### PR DESCRIPTION
**Pull request by KKcorps**
When user chooses to skip the time value check in table config, we need to gracefully handle the
segment startTime and endTime.
If the startTime and endTime for the segment are simply not within the valid time range, we set them to
current timestamps.
If they are within the valid time range, we check if the endTime falls outside retention time range.
In case it falls outside the time range, we replace endTime with currentTime so that the segments don't
get instantly deleted by retention manager after getting uploaded.

### CODE REVIEW

**Comment by KKcorps** [comment]
For now, only moved the code from previous PR (#9355) . 
Last comments by @Jackie-Jiang 

> We don't want to change the start/end time within the segment metadata, instead we want to enhance the time management code to gracefully handle the segments with invalid time values.
> Suggest using a separate PR for the time management change, and only keep the sample value check in this PR.

The new flag `skipSegmentTimeCheck: true` allows ingestion to not fail because of timestamp issues. As far as updating the timestamps in metadata is concerned, we need to find a way for users to specify the correct time-formats. The issue is we don't allow backward-incompatible change in the schema so if an user fixes the segment metadata timestamp via some new API, we won't be able to reflect those changes in the schema. This inconsistency may break a lot of things.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java [hunk 1]
```diff
@@ -765,6 +766,28 @@ private void writeMetadata()
...
+            endTime = now;
+          }
+          SegmentsValidationAndRetentionConfig validationConfig = _config.getTableConfig().getValidationConfig();
+          String retentionTimeUnitConfig = validationConfig.getRetentionTimeUnit();
+          String retentionTimeValueConfig = validationConfig.getRetentionTimeValue();
+          try {
+            TimeUnit retentionTimeUnit = TimeUnit.valueOf(retentionTimeUnitConfig.toUpperCase());
+            long retentionTimeValue = Long.parseLong(retentionTimeValueConfig);
+            long retentionTimeMs = retentionTimeUnit.toMillis(retentionTimeValue);
+            if (endTime <= (now - retentionTimeMs)) {
```
Comment (jackjlli) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:784]: nit: put some comments to explicitly explain the rationale of doing this here.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java [hunk 2]
```diff
@@ -765,6 +766,28 @@ private void writeMetadata()
...
+              endTime = now;
+            }
+          } catch (Exception e) {
+            LOGGER.warn("Invalid retention time: {} {} for table: {}, skip", retentionTimeUnitConfig,
```
Comment (jackjlli) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:788]: nit: it'd be good to state what step is skipped here.

**Review by Jackie-Jiang — COMMENTED** [review]
I feel this won't work properly for the following reasons:
1. The start/end time doesn't match the actual data, so the pruning logic will prune the segment unexpected
2. User might mistakenly push an old segment, and we don't want to keep the segment around

I feel the purpose for this PR is to prevent user putting wrong time format/unit in the schema. If we detect that the time value does not match the time format/unit, or the converted timestamp is not in valid range, we can put start/end time as `null`, and also store the raw min/max time in the ZK metadata. Essentially the idea in #9413

**Comment by KKcorps** [comment]
Yeah, I will go ahead with the idea of putting nulls in the metadata. This logic was created before we decided on `fixAPI`. Closing this PR.

**Pull request closed without merging**